src/Type.ml

```diff
@@ -1,23 +1,36 @@
 open Exceptions
+open Core
 
 type t = INT
        | BOOL
        | CHAR
        | STRING
        | LIST
-       [@@deriving sexp]
+       | TVAR of string
+       | ARRAY of (t * t)
+       [@@deriving compare,sexp]
 
-let of_string : string -> t = function
+let of_atomic_string (s:string) : t =
+  match s with 
   | "Int"    -> INT
   | "Bool"   -> BOOL
   | "Char"   -> CHAR
   | "String" -> STRING
   | "List"   -> LIST
-  | s -> raise (Parse_Exn ("Could not parse type `" ^ s ^ "`."))
+  | _ -> raise (Parse_Exn ("Could not parse type `" ^ s ^ "`."))
 
-let to_string : t -> string = function
+let rec of_string (sexp: Sexp.t) : t =
+  match sexp with
+    | Sexp.(Atom v) -> (of_atomic_string v)
+    | Sexp.(List ([Atom("Array");index;value])) -> ARRAY ((of_string (index)),(of_string (value)))
+    | _ -> raise (Parse_Exn ("Could not parse type `" ^ (Sexp.to_string_hum sexp) ^ "`."))
```

Comment:
This function should be named `of_sexp`. If we actually still need the `of_string` function, then you could insert the Sexp parsing call there and then call `of_sexp`.

```ocaml
let rec of_sexp (sexp: Sexp.t) : t = let open Sexp in function
  | Atom "Int"    -> INT
  | Atom "Bool"   -> BOOL
  | Atom "Char"   -> CHAR
  | Atom "String" -> STRING
  | Atom "List"   -> LIST
  | List [Atom "Array" ; index ; value]
    -> ARRAY ((of_sexp index) , (of_sexp value))
  | s -> raise (Parse_Exn ("Could not parse type `" ^ (Sexp.to_string_hum s) ^ "`."))
```